mkosi/run.py

```diff
@@ -244,72 +249,97 @@ def preexec() -> None:
             # expect it to pick.
             assert nfd == SD_LISTEN_FDS_START + i
 
-    # First, check if the sandbox works at all before executing the command.
-    if sandbox and (rc := subprocess.run(sandbox + ["true"]).returncode) != 0:
-        log_process_failure(sandbox, cmdline, rc)
-        raise subprocess.CalledProcessError(rc, sandbox + cmdline)
+    with sandbox as sbx:
+        prefix = [os.fspath(x) for x in sbx]
 
-    if subprocess.run(sandbox + ["sh", "-c", f"command -v {cmdline[0]}"], stdout=subprocess.DEVNULL).returncode != 0:
-        die(f"{cmdline[0]} not found.", hint=f"Is {cmdline[0]} installed on the host system?")
+        # First, check if the sandbox works at all before executing the command.
+        if prefix and (rc := subprocess.run(prefix + ["true"]).returncode) != 0:
+            log_process_failure(prefix, cmdline, rc)
+            raise subprocess.CalledProcessError(rc, prefix + cmdline)
 
-    if (
-        foreground and
-        sandbox and
-        subprocess.run(sandbox + ["sh", "-c", "command -v setpgid"], stdout=subprocess.DEVNULL).returncode == 0
-    ):
-        sandbox += ["setpgid", "--foreground", "--"]
+        if subprocess.run(
+            prefix + ["sh", "-c", f"command -v {cmdline[0]}"],
+            stdout=subprocess.DEVNULL,
+        ).returncode != 0:
+            die(f"{cmdline[0]} not found.", hint=f"Is {cmdline[0]} installed on the host system?")
 
-    if pass_fds:
-        # We don't know the PID before we start the process and we can't modify the environment in preexec_fn so we
-        # have to spawn a temporary shell to set the necessary environment variables before spawning the actual
-        # command.
-        sandbox += ["sh", "-c", f"LISTEN_FDS={len(pass_fds)} LISTEN_PID=$$ exec $0 \"$@\""]
+        if (
+            foreground and
+            prefix and
+            subprocess.run(prefix + ["sh", "-c", "command -v setpgid"], stdout=subprocess.DEVNULL).returncode == 0
+        ):
+            prefix += ["setpgid", "--foreground", "--"]
+
+        if pass_fds:
+            # We don't know the PID before we start the process and we can't modify the environment in preexec_fn so we
+            # have to spawn a temporary shell to set the necessary environment variables before spawning the actual
+            # command.
+            prefix += ["sh", "-c", f"LISTEN_FDS={len(pass_fds)} LISTEN_PID=$$ exec $0 \"$@\""]
+
+        if prefix and innerpid:
+            r, w = os.pipe2(os.O_CLOEXEC)
+            q = fcntl.fcntl(w, fcntl.F_DUPFD_CLOEXEC, len(pass_fds) + 1)
```

Comment:
This is to make the fd the right number for pass_fds, right? If so, a comment to that extent would make it easier to read.